src/preact/use-in-view.js

```diff
@@ -0,0 +1,48 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {useEffect, useLayoutEffect, useRef, useState} from './index';
+
+/**
+ * @param {HTMLElement} node
+ * @param {function():(function():undefined|undefined)} effect
+ * @param {!Array<*>=} opt_deps
+ * @return {object} ref
+ */
+export function useInViewEffect(node, effect, opt_deps) {
+  const {0: isIntersecting, 1: set} = useState(true);
+  const observerRef = useRef(null);
+  if (observerRef.current === null) {
+    observerRef.current = new IntersectionObserver(entries => {
+      const last = entries[entries.length - 1];
+      set(last.isIntersecting);
+    });
+  }
+
+  useLayoutEffect(() => {
+    const {current: observer} = observerRef;
+    if (node) {
+      observer.observe(node);
+    }
+    return () => observer.disconnect();
+  }, [node, observerRef.current]);
+
+  useEffect(() => {
+    if (isIntersecting) {
+      effect();
```

Comment:
Canonically this should be `return effect()`

Reply:
The `return` of `useEffect` takes a [cleanup callback](https://reactjs.org/docs/hooks-reference.html#cleaning-up-an-effect)--why does it make sense to return in this case?

Reply:
I think the cleanup callbacks are an absolute must and an expectation for effects. For instance, consider you have a use cases where you want to run the `setInterval` frequently while an element is in the viewport. The cleanup callback solves this for us elegantly.

```
useInViewEffect(ref, () => {
  const handle = setInterval(() => {
    // Do something expensive and often.
  }, 100);
  return () => clearInterval(handle);
}, [...])
```